src/Browser/Test/LegacyExtension.php

```diff
@@ -0,0 +1,134 @@
+ */
+
+namespace Zenstruck\Browser\Test;
+
+use Zenstruck\Browser;
+
+/**
+ * @author Kevin Bond <kevinbond@gmail.com>
+ */
+class LegacyExtension
+{
+    /** @var Browser[] */
+    private static array $registeredBrowsers = [];
+    private static bool $enabled = false;
+
+    /** @var array<string,array<string,string[]>> */
+    private array $savedArtifacts = [];
+
+    /**
+     * @internal
+     */
+    public static function registerBrowser(Browser $browser): void
+    {
+        if (!self::$enabled) {
+            return;
+        }
+
+        self::$registeredBrowsers[] = $browser;
+    }
+
+    public function executeBeforeFirstTest(): void
+    {
+        self::$enabled = true;
+    }
+
+    public function executeBeforeTest(string $test): void
+    {
+        self::reset();
+    }
+
+    public function executeAfterTest(string $test, float $time): void
+    {
+        foreach (self::$registeredBrowsers as $browser) {
+            foreach ($browser->savedArtifacts() as $category => $artifacts) {
+                if (!\count($artifacts)) {
+                    continue;
+                }
+
+                $this->savedArtifacts[$test][$category] = $artifacts;
+            }
+        }
+
+        self::reset();
+    }
+
+    public function executeAfterLastTest(): void
+    {
+        if (empty($this->savedArtifacts)) {
+            return;
+        }
+
+        echo "\n\nSaved Browser Artifacts:";
+
+        foreach ($this->savedArtifacts as $test => $categories) {
+            echo "\n\n  {$test}";
+
+            foreach ($categories as $category => $artifacts) {
+                echo "\n    {$category}:";
+
+                foreach ($artifacts as $artifact) {
+                    echo "\n      * {$artifact}:";
+                }
+            }
+        }
+    }
+
+    public function executeAfterTestError(string $test, string $message, float $time): void
+    {
+        self::saveBrowserStates($test, 'error');
+    }
+
+    public function executeAfterTestFailure(string $test, string $message, float $time): void
+    {
+        self::saveBrowserStates($test, 'failure');
+    }
+
+    private static function saveBrowserStates(string $test, string $type): void
+    {
+        if (empty(self::$registeredBrowsers)) {
+            return;
+        }
+
+        $filename = \sprintf('%s_%s', $type, self::normalizeTestName($test));
+
+        foreach (self::$registeredBrowsers as $i => $browser) {
+            try {
+                $browser->saveCurrentState("{$filename}__{$i}");
+            } catch (\Throwable $e) {
+                // noop - swallow exceptions related to dumping the current state so as to not
+                // lose the actual error/failure.
+            }
+        }
+    }
+
+    private static function normalizeTestName(string $name): string
+    {
+        // Try to match for a numeric data set index. If it didn't, match for a string one.
+        if (!\preg_match('#^([\w:\\\]+)(.+\#(\d+))#', $name, $matches)) {
```

Comment:
it needed a slight adapt here, it was `'#^([\w:\\\]+)(.+\#(\d+).+)?$#` & `'#^([\w:\\\]+)(.+"([\w ]+)".+)?$#` to extract names like `Namespace\TestClass::testMethod with data set #0 ('test data')` / `data set "set name" (test data)`.
Under phpunit 10 it returns names without "(test data)" (like ``Namespace\TestClass::testMethod with data set #0` / `data set "set name"`)

Reply:
A test should be written to check if both versions (above and below phpunit 10 ) work fine?

Reply:
@nikophil indeed, I tested it a bit and there's a side effect : the case without datasets (`Namespace\TestClass::testMethod` for example) that isn't captured by this...
I put the PR as draft before finding the solution ...

Are those the possible cases to capture or are there more ? : 

- `Namespace\TestClass::testMethod`
- `Namespace\TestClass::testMethod with data set #0 (data)`
- `Namespace\TestClass::testMethod with data set "test" (data)`
- `Namespace\TestClass::testMethod with data set #0` ( : new in phpunit 10)
- `Namespace\TestClass::testMethod with data set "test"`

Reply:
since the support for phpunit 10 test outputs, without drops on previous version's formats, seems ok here (not totaly tested the edge cases mentioned in #67 but doesn't seem the topic here), can I resolve this ?